lib/SingleEventManager.lua

```diff
@@ -14,132 +14,145 @@
+	return self
 end
 
-function SingleEventManager.new()
-	local self = {}
+function SingleEventManager:connect(instance, key, method)
+	self:_connectInternal(instance, instance[key], method)
+end
 
-	self._hookCache = {}
+function SingleEventManager:connectProperty(instance, key, method)
+	self:_connectInternal(instance, instance:GetPropertyChangedSignal(key), method)
+end
 
-	setmetatable(self, SingleEventManager)
+--[[
+	Disconnects the hook attached to the event named `key` on the given
+	`instance` if there is one, otherwise does nothing.
 
-	return self
+	Note that `key` must identify a valid property on `instance`, or this method
+	will throw.
+]]
+function SingleEventManager:disconnect(instance, key)
+	self:_disconnectInternal(instance, instance[key])
 end
 
-function SingleEventManager:connect(rbx, key, method)
-	local rbxHooks = self._hookCache[rbx]
+--[[
+	Disconnects the hook attached to the property changed signal on `instance`
+	with the name `key` if there is one, otherwise does nothing.
 
-	if rbxHooks then
-		local existingHook = rbxHooks[key]
+	Note that `key` must identify a valid property on `instance`, or this method
+	will throw.
+]]
+function SingleEventManager:disconnectProperty(instance, key)
+	self:_disconnectInternal(instance, instance:GetPropertyChangedSignal(key))
+end
 
-		if existingHook then
-			if existingHook.method == method then
-				return
-			end
+--[[
+	Disconnects any hooks managed by SingleEventManager associated with
+	`instance`.
 
-			existingHook.connection:Disconnect()
-		end
+	Calling disconnectAll with an untracked instance won't do anything.
+]]
+function SingleEventManager:disconnectAll(instance)
+	local instanceHooks = self._hooks[instance]
 
-		rbxHooks[key] = createHook(rbx, key, method)
-	else
-		rbxHooks = {}
-		rbxHooks[key] = createHook(rbx, key, method)
+	if instanceHooks == nil then
+		return
+	end
 
-		self._hookCache[rbx] = rbxHooks
+	for _, hook in pairs(instanceHooks) do
+		hook.connection:Disconnect()
 	end
+
+	self._hooks[instance] = nil
 end
 
-function SingleEventManager:connectProperty(rbx, key, method)
-	local rbxHooks = self._hookCache[rbx]
-	local formattedKey = formatChangeKey(key)
+--[[
+	Creates a hook using the given event and method and associates it with the
+	given instance.
 
-	if rbxHooks then
-		local existingHook = rbxHooks[formattedKey]
+	Generally, `event` should directly associated with `instance`, but that's
+	unchecked in this code.
+]]
+function SingleEventManager:_connectInternal(instance, event, method)
+	local instanceHooks = self._hooks[instance]
 
-		if existingHook then
-			if existingHook.method == method then
-				return
-			end
+	if instanceHooks == nil then
+		instanceHooks = {}
+		self._hooks[instance] = instanceHooks
+	end
 
-			existingHook.connection:Disconnect()
-		end
+	local existingHook = instanceHooks[event]
 
-		rbxHooks[formattedKey] = createChangeHook(rbx, key, method)
+	if existingHook ~= nil then
+		existingHook.method = method
```

Comment:
This is *nice*. I wonder how much this speeds up changing event connections in the case where they're defined in `render`?

Reply:
When I did some unscientific napkin benchmarking, it's somewhere around 100x faster in the worst case. Not sure if it was ever a bottleneck, but it's interesting!